R/GeospatialFunctions.R

```diff
@@ -203,44 +229,303 @@ fetchATTAINS <- function(.data) {
+    other_features <- baseurls[2:4] %>%
+      purrr::map(~ feature_downloader(baseurls = ., sf_bbox = bbox))
+    
+    final_features <- list("ATTAINS_catchments" = catchment_features, 
+                           "ATTAINS_points" = other_features[[1]], 
+                           "ATTAINS_lines" = other_features[[2]], 
+                           "ATTAINS_polygons" = other_features[[3]])
+    
+    return(final_features)
+  }
+}
 
-  final_features <- baseurls %>%
-    purrr::map(~ feature_downloader(.))
-
-  names(final_features) <- c("ATTAINS_catchments", "ATTAINS_points", "ATTAINS_lines", "ATTAINS_polygons")
 
-  return(final_features)
+#' fetchNHD
+#'
+#' Fetches NHD catchment features from either the high resolution or medium resolution version of the National Hydrography Dataset (NHD) that intersect TADA Water Quality Portal observations.
+#'
+#' @param .data A dataframe created by `TADA_DataRetrieval()` or the sf equivalent made by `TADA_MakeSpatial()`.
+#' @param resolution Whether to download the high resolution ("NHDPlus HiRes") or medium resolution ("NHDPlus V2") version of the National Hydrography Dataset (NHD). Default is "NHDPlus HiRes".
+#'
+#' @return A list containing all flowline, waterbody, and catchment features that intersect the WQP observations of interest.
+#'
+#' @seealso [TADA_DataRetrieval()]
+#' @seealso [TADA_MakeSpatial()]
+#'
+#' @examples
+#' \dontrun{
+#' tada_data <- TADA_DataRetrieval(
+#'   startDate = "1990-01-01",
+#'   endDate = "1995-12-31",
+#'   characteristicName = "pH",
+#'   statecode = "RI",
+#'   applyautoclean = TRUE
+#' )
+#'
+#' attains_data <- fetchNHD(.data = tada_data, resolution = "NHDPlusV2")
+#' }
+fetchNHD <- function(.data, resolution = "NHDPlus HiRes"){
+  
+  suppressMessages(suppressWarnings({
+    
+    sf::sf_use_s2(FALSE)
+    # If data is already spatial, just make sure it is in the right CRS
+    # and add unique WQP ID for identifying obs with more than one ATTAINS assessment unit
+    
+    if (!is.null(.data) & inherits(.data, "sf")) {
+      if (sf::st_crs(.data)$epsg != 4326) {
+        geospatial_data <- .data %>%
+          sf::st_transform(4326) #%>%
+        #tibble::rowid_to_column(var = "index")
+      } else {
+        geospatial_data <- .data
+      }
+    } else {
+      # ... Otherwise transform into a spatial object then do the same thing:
+      geospatial_data <- .data %>%
+        # convert dataframe to a spatial object
+        TADA_MakeSpatial(.data = ., crs = 4326) %>%
+        # add unique WQP ID for identifying obs with more than one ATTAINS assessment unit
+        #tibble::rowid_to_column(var = "index") %>%
+        dplyr::mutate(geometry_join = geometry)
+    }
+    
+  }))
+  
+  # Reduce WQP data to unique coordinates 
+  unique_sites <- dplyr::distinct(geospatial_data, geometry)
+  
+  # If user wants HighRes NHD...
+  if(resolution %in% c("NHDPlus HiRes", "NHDPlus HR", "HighRes", "NHDHR", "HR", "NHD HR")){
```

Comment:
question: Are these all potential inputs to the resolution argument?

Reply:
changed to Hi and Med. Error handling included!